### PR TITLE
Allow setting 'rel' and 'title' attributes for stylesheets.

### PR DESCRIPTION
**Pull request by Jellby**
This addresses #1764. I added documentation too, but no test (I'm not sure how that works). Please tell me if there is something else wrong or missing.

### CODE REVIEW

File: sphinx/application.py
Comment (lehmannro) [inline sphinx/application.py]: I don't think passing arbitrary strings to `rel` is good API design.  This should be some kind of enum or, which I think would be totally sufficient for the use cases at hand, just an `alternate=bool` flag.

Comment (Jellby) [inline sphinx/application.py]: Better with the new commit?

**Comment by Jellby** [comment]
How can I create a test? The [Developer's Guide](http://sphinx-doc.org/latest/devguide.html) says "Write a test", but how do I write it and how do I specify what the correct result is? In this case I need to test that the add_stylesheet method accepts the two additional arguments and that setting the css_files in templates works as intended (i.e., check the html builder with custom conf.py and modified template).

EDIT: Nevermind, I think I figured it out (see next commit).

File: sphinx/themes/basic/layout.html
Comment (shimizukawa) [inline sphinx/themes/basic/layout.html]: This logic shouldn't be wrote on html template because it makes difficult to read and/or understand for developers and users. LGTM except this lines.

Comment (Jellby) [inline sphinx/themes/basic/layout.html]: But it has to be in the template if I want to allow css_files being modified in a template file (see tests/roots/test-stylesheets/_templates/layout.html). Or can I process css_files between its definition with:

```
{% set css_files = ... %}
```

and its implicit use with:

```
{{ css() }}
```

?

The problem is that the default value for "alternate", if it's not specified, depends on whether or not a "title" attribute was provided. I can easily handle this on the python side (in add_stylesheet), but that's not the only way to add css files.

Comment (shimizukawa) [inline sphinx/themes/basic/layout.html]: IMO, css_files variable must not have multiple structure according to circumstances because it is difficult for sphinx users to understand and to tweak template. And also it is hard for maintainers to keep it stable.

BTW, I reconsidered about this change. Then I can't merge this for 1.4 release because side effect of this change is large for users who are using custom template that references to `css_files`. To resolve this concern, we should provide another variable.

**Comment by Jellby** [comment]
@shimizukawa I modified it to use a separate variable for the properties. This should make it completely backwards-compatible. The logic in the template is also slightly clearer now, I believe, although I don't like the way to update a dict in jinja.

File: sphinx/themes/basic/layout.html
Comment (shimizukawa) [inline sphinx/themes/basic/layout.html]: `css_props` is nice, however this if/else/set structure is too complicated to read.
My idea I mentioned is like this:

```
    {#- Process the custom css files, with their alternate and title properties if available #}
    {%- for cssfile in css_files %}
    <link rel="{{ css_props[cssfile].rel }}" href="{{ pathto(cssfile, 1) }}" type="text/css" title="{{ css_props[cssfile].csstitle }}" />
    {%- endfor %}
```

Trade-off is, `title=""` will be rendered if title is not provided.
@Jellby how about this?

File: tests/roots/test-stylesheets/_templates/layout.html
Comment (shimizukawa) [inline tests/roots/test-stylesheets/_templates/layout.html]: I don't think the functionality is needed to modify title and alternate attribute from template side.

File: sphinx/application.py
Comment (shimizukawa) [inline sphinx/application.py]: default value of alternate must be `False`

File: sphinx/application.py
Comment (shimizukawa) [inline sphinx/application.py]: Sometime 'alternate' and 'title' keys are given, sometime not.
I think `props = {'alternate': alternate, 'title': title}` is simple and better.

File: doc/templating.rst
Comment (shimizukawa) [inline doc/templating.rst]: IMO, rewriting a template variable is bad manner.
Even if you or someone want to use such workaround, official document should not provide such information.

File: doc/templating.rst
Comment (shimizukawa) [inline doc/templating.rst]: same as above.

However, `css_props` data structure information is useful for everyone.
Can you update please?

**Review by shimizukawa — CHANGES_REQUESTED** [review]
Reviewed. @Jellby please take a look.

File: sphinx/themes/basic/layout.html
Comment (Jellby) [reply sphinx/themes/basic/layout.html]: I'm afraid that won't work. In HTML/CSS a stylesheet with a `title` attribute is treated differently from another without it. A stylesheet without `title` is always applied (persistent). A stylesheet with `title` is applied by default, but is disabled if another alternate stylesheet is selected (preferred). See https://developer.mozilla.org/en/docs/Correctly_Using_Titles_With_External_Stylesheets for example. Your suggestion would make it impossible to specify persistent stylesheets (unless a browser considers an empty title as no title, but I haven't found anything in the specification indicating this should be the case).

And then, I would like to allow setting the properties in a template file, not only through the `add_stylesheet` method, I believe it can be useful or needed in some cases, probably when designing a theme. If the code is difficult to read, would it help if I add some comments explaining its purpose?

File: tests/roots/test-stylesheets/_templates/layout.html
Comment (Jellby) [reply tests/roots/test-stylesheets/_templates/layout.html]: As pointed out above, what if I want to set that in a theme? Can a theme use python code?

File: sphinx/application.py
Comment (Jellby) [reply sphinx/application.py]: That could be. Personally, I'd say it is more often that one wants to set a stylesheet with title as alternate, so I wanted to make the "default" True if there is a title and False if there isn't. To get this I had to store somehow whether the alternate argument had been given or not: basically, to pass the unprocessed information to the template (see above).

File: sphinx/application.py
Comment (Jellby) [reply sphinx/application.py]: If I do that, and one doesn't pass a title, would e.g. `{% if css_props[cssfile].title is defined %}` work in the template? From a python point of view, it is defined (as `None`).

File: doc/templating.rst
Comment (Jellby) [reply doc/templating.rst]: But it's already there for `script_files`...

**Comment by Jellby** [comment]
@shimizukawa Would you please reconsider your objections, or reply? The HTML logic for alternate stylesheets is a bit involved, and I really want to set alternate stylesheets in a template file, as part of a theme.

File: doc/templating.rst
Comment (tk0miya) [reply doc/templating.rst]: +1 for shimizukawa.
I feel this is hacky approach too. So it should not be described in our document.
I think our templates around `script_files` should be also fixed.

To do that, it would be nice to add new extensible block like following:
```
{%- block css %}
  {{ super() }}
  <link rel="stylesheet" href="_static/mystyle.css" type="text/css" />
{%- endblock %}
```

File: sphinx/themes/basic/layout.html
Comment (tk0miya) [reply sphinx/themes/basic/layout.html]: Agreed, `title=""` is not good. But I also agreed to this is too complicated.

I prefer to pack the CSS information to single object in python side.
For example:
```
{%- for css in css_files %}
<link rel="{{ css.rel }}" href="{{ pathto(css.filename, 1) }}" type="text/css"{% if css.title is none %} title="{{ css.title }}"{% endif %} />
<%- endfor %}
```

How about this?

**Review by tk0miya — CHANGES_REQUESTED** [review]

**Comment by tk0miya** [comment]
+1 for basic idea, but we need to arrange the code.

**Comment by Jellby** [comment]
@shimizukawa @tk0miya That suggestion about the template block and the reference to `super()` solved all my problems. I can now get rid of all the complicate template logic and assume that if anyone (like myself) wants to modify the template, they can write HTML code directly. After that, I could also implement the other suggestions/requests. I added a ``scripts`` block too, I figured that doesn't harm.

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
LGTM! Thank you for update :-)

**Comment by tk0miya** [comment]
ping @shimizukawa 
Could you review this please? It's time to v1.6!

File: sphinx/themes/basic/layout.html
Comment (shimizukawa) [inline sphinx/themes/basic/layout.html:109]: This is a incompatible change.
I think it would be better if `css` object (current implementation is using dict) has `__str__` and return a value of `filename` attribute to keep compatibility.
@tk0miya How do you think?

**Review by shimizukawa — APPROVED** [review]
Reviewed. Thanks @Jellby!

@tk0miya I leave a comment for about incompatibility.

File: sphinx/themes/basic/layout.html
Comment (tk0miya) [reply sphinx/themes/basic/layout.html:109]: Yes, you're right. I'll arrange that after merging.

**Comment by tk0miya** [comment]
@Jellby Thank you for long-term and tough contributing! I'm glad to merge this :-)

**Pull request merged**
